cfg.go

```diff
@@ -148,96 +159,140 @@ type Node interface {
+func (b *Builder) NextFreshVar() string {
+	b.FreshVarCnt++
+	return b.SameFreshVar()
 }
-*/
 
 func (b *Builder) BlockToNode(block *cfg.Block, pred Node) {
-	first := pred
-	currPred := pred
+	var blockFirst Node = nil //represents first node of this block
+	currLast := pred
 	for _, astNode := range block.Nodes {
-		f, l := b.astToNode(astNode, currPred)
-		if f != nil {
-			if first == pred {
-				first = f
+		first, last := b.astToNode(astNode)
+		if first != nil {
+			connect(currLast, first)
+			currLast = last
+			//if blockFirst is nil, that means this is first node of this block
+			if blockFirst == nil {
+				blockFirst = first
 			}
-			currPred = l
 		}
 	}
-	//don't store blocks, that had no nodes
-	if first != pred {
-		b.BlockNode[block] = first
+	//if current block hasn't created any new node, we skip this part
+	if blockFirst != nil {
+		b.BlockNode[block] = blockFirst
+		connect(pred, blockFirst)
 	}
 	for _, suc := range block.Succs {
 		n, ok := b.BlockNode[suc]
 		if ok {
-			connect(currPred, n)
+			connect(currLast, n)
 			continue
 		}
-		b.BlockToNode(suc, currPred)
+		b.BlockToNode(suc, currLast)
 	}
 }
 
-func (b *Builder) astToNode(a ast.Node, pred Node) (first, last Node) {
+//appends node to first-last sequence of nodes and returns a new first-last sequence
+func (b *Builder) appendNode(n Node, f Node, l Node) (Node, Node) {
+	if f == nil {
+		return n, n
+	}
+	connect(l, n)
+	return f, n
+}
+
+//returns first and last node created by this AST node
+func (b *Builder) astToNode(a ast.Node) (first, last Node) {
 	switch a := a.(type) {
 	case *ast.AssignStmt:
-		first, last = b.assignStmtToNode(a)
-	case *ast.ReturnStmt:
-	}
-	if first != nil {
-		connect(pred, first)
+		//extends current first-last sequence with new nodes and returns new first-last sequence
+		first, last = b.assignLhsToNode(a.Lhs[0], a.Rhs[0], first, last)
 	}
 	return first, last
 }
 
-func (b *Builder) assignStmtToNode(stmt *ast.AssignStmt) (first, last Node) {
-	switch lhs := stmt.Lhs[0].(type) { // TODO: multivariable
+//decomposes lhs to string
+func (b *Builder) assignLhsToNode(lhsExp ast.Expr, rhsExp ast.Expr, f Node, l Node) (first, last Node) {
+	switch lhs := lhsExp.(type) {
 	case *ast.Ident:
-		switch rhs := stmt.Rhs[0].(type) {
-		case *ast.Ident:
-			if rhs.Name == "nil" {
-				n := NewNullNode(lhs.Name, stmt)
-				return n, n
-			} else {
-				n := NewAssignNode(lhs.Name, rhs.Name, stmt)
-				return n, n
-			}
-		case *ast.UnaryExpr:
-			if rhs.Op == token.AND { //&
-				id, ok := rhs.X.(*ast.Ident)
-				if !ok {
-					panic("expected ident")
+		first, last = b.assignRhsToNode(lhs.Name, rhsExp, f, l)
+	case *ast.StarExpr:
+		//check if we need to normalize lhs StarExpr even more
+		switch id := lhs.X.(type) {
+		case *ast.Ident: //no need to normalize
+			//now we'll peek the rhs
+			switch rhs := rhsExp.(type) {
+			case *ast.Ident:
+				if rhs.Name != "nil" {
+					return b.appendNode(NewDerefNode(id.Name, rhs.Name, rhsExp), f, l)
```

Comment:
It could still be `nil` right? In cases like

```
var p *int
println(*p)
```

Or is it taken care of?

Reply:
Yeah, `p` can still be `nil`, but there's no way of knowing that until we run nilness analysis. This case is meant to represent explicit `nil` assign, like `p = nil` 😄

Reply:
Ah okay I thought about that, but just to be sure.